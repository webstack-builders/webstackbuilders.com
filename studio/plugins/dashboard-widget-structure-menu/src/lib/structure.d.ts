export declare function isSubscribable(thing: any): any;
export declare function isStructure(structure: any): any;
export declare function serializeStructure(item: any, context: any, resolverArgs?: never[]): any;
export declare function getDefaultStructure(): any;
export declare function loadStructure(): import("rxjs").Observable<unknown>;
//# sourceMappingURL=structure.d.ts.map