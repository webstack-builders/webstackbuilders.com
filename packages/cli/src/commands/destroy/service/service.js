import { getDefaultArgs } from '../../../lib'
import { builder, files } from '../../generate/service/service'
import { createYargsForComponentDestroy } from '../helpers'

// This function wraps files(), so we can pass templateVars. templateVars
// referenced in a file template must be defined, otherwise template rendering
// fails. This way we can pass stub values for templateVars and do not define
// fake builder flags for destroy command just to make templates work.
//
// Better solution would be to split file paths resolving and template
// rendering into separate functions. See more in this PR discussion:
// https://github.com/redwoodjs/redwood/pull/487#issue-411204396
export const filesWithTemplateVars = (templateVars) => {
  return (args) => files({ ...args, ...templateVars })
}

export const { command, description, handler, tasks } =
  createYargsForComponentDestroy({
    componentName: 'service',
    filesFn: filesWithTemplateVars({ ...getDefaultArgs(builder), crud: true }),
    shouldEnsureUniquePlural: true,
  })
