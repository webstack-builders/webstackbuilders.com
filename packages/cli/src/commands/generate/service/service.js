import camelcase from 'camelcase'
import pascalcase from 'pascalcase'
import pluralize from 'pluralize'
import terminalLink from 'terminal-link'

import { getSchema, transformTSToJS } from '../../../lib'
import { yargsDefaults } from '../../generate'
import {
  templateForComponentFile,
  createYargsForComponentGeneration,
} from '../helpers'

const DEFAULT_SCENARIO_NAMES = ['one', 'two']

// parses the schema into scalar fields, relations and an array of foreign keys
export const parseSchema = async (model) => {
  const schema = await getSchema(model)
  const relations = {}
  let foreignKeys = []

  // aggregate the plain String, Int and DateTime fields
  let scalarFields = schema.fields.filter((field) => {
    if (field.relationFromFields) {
      // only build relations for those that are required
      if (field.isRequired && field.relationFromFields.length !== 0) {
        relations[field.name] = {
          foreignKey: field.relationFromFields,
          type: field.type,
        }
      }
      foreignKeys = foreignKeys.concat(field.relationFromFields)
    }

    return (
      field.isRequired &&
      !field.hasDefaultValue && // don't include fields that the database will default
      !field.relationName // this field isn't a relation (ie. comment.post)
    )
  })

  return { scalarFields, relations, foreignKeys }
}

export const scenarioFieldValue = (field) => {
  const randFloat = Math.random() * 10000000
  const randInt = parseInt(Math.random() * 10000000)

  switch (field.type) {
    case 'Boolean':
      return true
    case 'DateTime':
      return new Date().toISOString().replace(/\.\d{3}/, '')
    case 'Decimal':
    case 'Float':
      return randFloat
    case 'Int':
      return randInt
    case 'Json':
      return { foo: 'bar' }
    case 'String':
      return field.isUnique ? `String${randInt}` : 'String'
    default: {
      if (field.kind === 'enum' && field.enumValues[0]) {
        return field.enumValues[0].dbName || field.enumValues[0].name
      }
    }
  }
}

export const fieldsToScenario = async (
  scalarFields,
  relations,
  foreignKeys
) => {
  const data = {}

  // remove foreign keys from scalars
  scalarFields.forEach((field) => {
    if (!foreignKeys.length || !foreignKeys.includes(field.name)) {
      data[field.name] = scenarioFieldValue(field)
    }
  })

  // add back in related models by name so they can be created with prisma create syntax
  for (const [relationName, relData] of Object.entries(relations)) {
    const relationModelName = relData.type
    const {
      scalarFields: relScalarFields,
      relations: relRelations,
      foreignKeys: relForeignKeys,
    } = await parseSchema(relationModelName)

    data[relationName] = {
      create: await fieldsToScenario(
        relScalarFields,
        relRelations,
        relForeignKeys
      ),
    }
  }

  return data
}

// creates the scenario data based on the data definitions in schema.prisma
export const buildScenario = async (model) => {
  const scenarioModelName = camelcase(model)
  const standardScenario = {
    [scenarioModelName]: {},
  }
  const { scalarFields, relations, foreignKeys } = await parseSchema(model)

  // turn scalar fields into actual scenario data
  for (const name of DEFAULT_SCENARIO_NAMES) {
    standardScenario[scenarioModelName][name] = await fieldsToScenario(
      scalarFields,
      relations,
      foreignKeys
    )
  }

  return standardScenario
}

// outputs fields necessary to create an object in the test file
export const fieldsToInput = async (model) => {
  const { scalarFields, foreignKeys } = await parseSchema(model)
  const modelName = camelcase(pluralize.singular(model))
  let inputObj = {}

  scalarFields.forEach((field) => {
    if (foreignKeys.includes(field.name)) {
      inputObj[field.name] = `scenario.${modelName}.two.${field.name}`
    } else {
      inputObj[field.name] = scenarioFieldValue(field)
    }
  })

  if (Object.keys(inputObj).length > 0) {
    return inputObj
  } else {
    return false
  }
}

// outputs fields necessary to update an object in the test file
export const fieldsToUpdate = async (model) => {
  const { scalarFields, relations, foreignKeys } = await parseSchema(model)
  const modelName = camelcase(pluralize.singular(model))
  let field, newValue, fieldName

  // find an editable scalar field, ideally one that isn't a foreign key
  field = scalarFields.find((scalar) => !foreignKeys.includes(scalar.name))

  // no non-foreign keys, so just take the first one
  if (!field) {
    field = scalarFields[0]
  }

  // if the model has no editable scalar fields, skip update test completely
  if (!field) {
    return false
  }

  if (foreignKeys.includes(field.name)) {
    // no scalar fields, change a relation field instead
    // { post: { foreignKey: [ 'postId' ], type: "Post" }, tag: { foreignKey: [ 'tagId' ], type: "Post" } }
    fieldName = Object.values(relations)[0].foreignKey
    newValue = `scenario.${modelName}.two.${field.name}`
  } else {
    fieldName = field.name

    // change scalar fields
    const value = scenarioFieldValue(field)
    newValue = value

    // depending on the field type, append/update the value to something different
    switch (field.type) {
      case 'Boolean': {
        newValue = !value
        break
      }
      case 'DateTime': {
        let date = new Date()
        date.setDate(date.getDate() + 1)
        newValue = date.toISOString().replace(/\.\d{3}/, '')
        break
      }
      case 'Decimal':
      case 'Float': {
        newValue = newValue + 1.1
        break
      }
      case 'Int': {
        newValue = newValue + 1
        break
      }
      case 'Json': {
        newValue = { foo: 'baz' }
        break
      }
      case 'String': {
        newValue = newValue + '2'
        break
      }
      default: {
        if (
          field.kind === 'enum' &&
          field.enumValues[field.enumValues.length - 1]
        ) {
          const enumVal = field.enumValues[field.enumValues.length - 1]
          newValue = enumVal.dbName || enumVal.name
        }
        break
      }
    }
  }

  return { [fieldName]: newValue }
}

export const files = async ({
  name,
  tests,
  relations,
  typescript,
  ...rest
}) => {
  const componentName = camelcase(pluralize(name))
  const model = pascalcase(pluralize.singular(name))
  const extension = 'ts'
  const serviceFile = templateForComponentFile({
    name,
    componentName: componentName,
    extension: `.${extension}`,
    apiPathSection: 'services',
    generator: 'service',
    templatePath: `service.${extension}.template`,
    templateVars: { relations: relations || [], ...rest },
  })
  const testFile = templateForComponentFile({
    name,
    componentName: componentName,
    extension: `.test.${extension}`,
    apiPathSection: 'services',
    generator: 'service',
    templatePath: `test.${extension}.template`,
    templateVars: {
      relations: relations || [],
      create: await fieldsToInput(model),
      update: await fieldsToUpdate(model),
      ...rest,
    },
  })
  const scenariosFile = templateForComponentFile({
    name,
    componentName: componentName,
    extension: `.scenarios.${extension}`,
    apiPathSection: 'services',
    generator: 'service',
    templatePath: `scenarios.${extension}.template`,
    templateVars: {
      scenario: await buildScenario(model),
      prismaTypeName: `${model}CreateArgs`,
      ...rest,
    },
  })

  const files = [serviceFile]
  if (tests) {
    files.push(testFile)
    files.push(scenariosFile)
  }

  // Returns
  // {
  //    "path/to/fileA": "<<<template>>>",
  //    "path/to/fileB": "<<<template>>>",
  // }
  return files.reduce((acc, [outputPath, content]) => {
    if (!typescript) {
      content = transformTSToJS(outputPath, content)
      outputPath = outputPath.replace('.ts', '.js')
    }

    return {
      [outputPath]: content,
      ...acc,
    }
  }, {})
}

export const defaults = {
  ...yargsDefaults,
  tests: {
    description: 'Generate test files',
    type: 'boolean',
  },
  crud: {
    default: false,
    description: 'Create CRUD functions',
    type: 'boolean',
  },
}

export const builder = (yargs) => {
  yargs
    .positional('name', {
      description: 'Name of the service',
      type: 'string',
    })
    .epilogue(
      `Also see the ${terminalLink(
        'Redwood CLI Reference',
        'https://redwoodjs.com/reference/command-line-interface#generate-service'
      )}`
    )
  Object.entries(defaults).forEach(([option, config]) => {
    yargs.option(option, config)
  })
}

export const { command, description, handler } =
  createYargsForComponentGeneration({
    componentName: 'service',
    filesFn: files,
    shouldEnsureUniquePlural: true,
  })
