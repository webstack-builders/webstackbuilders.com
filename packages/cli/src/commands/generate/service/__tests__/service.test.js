global.__dirname = __dirname
import path from 'path'

// Load mocks
import '../../../../lib/test'

import { getDefaultArgs } from '../../../../lib'
import * as service from '../service'

const extensionForBaseArgs = (baseArgs) =>
  baseArgs && baseArgs.typescript ? 'ts' : 'js'

const itReturnsExactly3Files = (baseArgs) => {
  test('returns exactly 3 files', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'User',
    })

    expect(Object.keys(files).length).toEqual(3)
  })
}
const itCreatesASingleWordServiceFile = (baseArgs) => {
  test('creates a single word service file', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'User',
    })
    const extension = extensionForBaseArgs(baseArgs)

    expect(
      files[
        path.normalize(
          `/path/to/project/api/src/services/users/users.${extension}`
        )
      ]
    ).toMatchSnapshot()
  })
}
const itCreatesASingleWordServiceTestFile = (baseArgs) => {
  test('creates a single word service test file', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'User',
    })
    const extension = extensionForBaseArgs(baseArgs)

    expect(
      files[
        path.normalize(
          `/path/to/project/api/src/services/users/users.test.${extension}`
        )
      ]
    ).toMatchSnapshot()
  })
}

const itCreatesASingleWordServiceScenarioFile = (baseArgs) => {
  test('creates a single word service scenario file', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'User',
    })
    const extension = extensionForBaseArgs(baseArgs)
    const filePath = path.normalize(
      `/path/to/project/api/src/services/users/users.scenarios.${extension}`
    )

    expect(Object.keys(files)).toContain(filePath)
  })
}

const itCreatesAMultiWordServiceFile = (baseArgs) => {
  test('creates a multi word service file', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'UserProfile',
    })
    const extension = extensionForBaseArgs(baseArgs)

    expect(
      files[
        path.normalize(
          `/path/to/project/api/src/services/userProfiles/userProfiles.${extension}`
        )
      ]
    ).toMatchSnapshot()
  })
}

const itCreatesAMultiWordServiceTestFile = (baseArgs) => {
  test('creates a multi word service test file', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'UserProfile',
    })
    const extension = extensionForBaseArgs(baseArgs)

    expect(
      files[
        path.normalize(
          `/path/to/project/api/src/services/userProfiles/userProfiles.test.${extension}`
        )
      ]
    ).toMatchSnapshot()
  })
}

const itCreatesASingleWordServiceFileWithCRUDActions = (baseArgs) => {
  test('creates a single word service file with CRUD actions', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'Post',
      crud: true,
    })
    const extension = extensionForBaseArgs(baseArgs)

    expect(
      files[
        path.normalize(
          `/path/to/project/api/src/services/posts/posts.${extension}`
        )
      ]
    ).toMatchSnapshot()

    // TODO
    // Mock Date, so we can take snapshots of tests and scenarios
  })
}

const itCreatesASingleWordServiceTestFileWithCRUDActions = (baseArgs) => {
  test('creates a service test file with CRUD actions', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'Post',
      crud: true,
    })
    const extension = extensionForBaseArgs(baseArgs)
    const filePath = path.normalize(
      `/path/to/project/api/src/services/posts/posts.test.${extension}`
    )

    expect(Object.keys(files)).toContain(filePath)
  })
}

const itCreatesAMultiWordServiceFileWithCRUDActions = (baseArgs) => {
  test('creates a multi word service file with CRUD actions', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'UserProfile',
      crud: true,
    })
    const extension = extensionForBaseArgs(baseArgs)
    const filePath = path.normalize(
      `/path/to/project/api/src/services/userProfiles/userProfiles.${extension}`
    )

    expect(Object.keys(files)).toContain(filePath)
  })
}

const itCreatesAMultiWordServiceTestFileWithCRUDActions = (baseArgs) => {
  test('creates a multi word service test file with CRUD actions', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'UserProfile',
      crud: true,
    })
    const extension = extensionForBaseArgs(baseArgs)
    const filePath = path.normalize(
      `/path/to/project/api/src/services/userProfiles/userProfiles.test.${extension}`
    )

    expect(Object.keys(files)).toContain(filePath)
  })
}

const itCreatesASingleWordServiceFileWithAHasManyRelation = (baseArgs) => {
  test('creates a single word service file with a hasMany relation', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'User',
      relations: ['userProfiles'],
    })
    const extension = extensionForBaseArgs(baseArgs)

    expect(
      files[
        path.normalize(
          `/path/to/project/api/src/services/users/users.${extension}`
        )
      ]
    ).toMatchSnapshot()
  })
}

const itCreatesASingleWordServiceFileWithABelongsToRelation = (baseArgs) => {
  test('creates a single word service file with a belongsTo relation', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'User',
      relations: ['identity'],
    })
    const extension = extensionForBaseArgs(baseArgs)

    expect(
      files[
        path.normalize(
          `/path/to/project/api/src/services/users/users.${extension}`
        )
      ]
    ).toMatchSnapshot()
  })
}

const itCreatesASingleWordServiceFileWithMultipleRelations = (baseArgs) => {
  test('creates a single word service file with multiple relations', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'User',
      relations: ['userProfiles', 'identity'],
    })
    const extension = extensionForBaseArgs(baseArgs)

    expect(
      files[
        path.normalize(
          `/path/to/project/api/src/services/users/users.${extension}`
        )
      ]
    ).toMatchSnapshot()
  })
}

const itCreatesAMultiWordServiceTestFileWithCRUDAndOnlyForeignKeyRequired = (
  baseArgs
) => {
  test('creates a multi word service test file with crud actions and only foreign as mandatory field', async () => {
    const files = await service.files({
      ...baseArgs,
      name: 'Transaction',
      crud: true,
    })
    const extension = extensionForBaseArgs(baseArgs)

    expect(
      files[
        path.normalize(
          `/path/to/project/api/src/services/transactions/transactions.test.${extension}`
        )
      ]
    ).toMatchSnapshot()
  })
}

describe('in javascript mode', () => {
  const baseArgs = { ...getDefaultArgs(service.defaults), tests: true }

  itReturnsExactly3Files(baseArgs)
  itCreatesASingleWordServiceFile(baseArgs)
  itCreatesASingleWordServiceTestFile(baseArgs)
  itCreatesASingleWordServiceScenarioFile(baseArgs)
  itCreatesAMultiWordServiceFile(baseArgs)
  itCreatesAMultiWordServiceTestFile(baseArgs)
  itCreatesASingleWordServiceFileWithCRUDActions(baseArgs)
  itCreatesASingleWordServiceTestFileWithCRUDActions(baseArgs)
  itCreatesAMultiWordServiceFileWithCRUDActions(baseArgs)
  itCreatesAMultiWordServiceTestFileWithCRUDActions(baseArgs)
  itCreatesASingleWordServiceFileWithAHasManyRelation(baseArgs)
  itCreatesASingleWordServiceFileWithABelongsToRelation(baseArgs)
  itCreatesASingleWordServiceFileWithMultipleRelations(baseArgs)
  itCreatesAMultiWordServiceTestFileWithCRUDAndOnlyForeignKeyRequired(baseArgs)
})

describe('in typescript mode', () => {
  const baseArgs = {
    ...getDefaultArgs(service.defaults),
    typescript: true,
    tests: true,
  }

  itReturnsExactly3Files(baseArgs)
  itCreatesASingleWordServiceFile(baseArgs)
  itCreatesASingleWordServiceTestFile(baseArgs)
  itCreatesASingleWordServiceScenarioFile(baseArgs)
  itCreatesAMultiWordServiceFile(baseArgs)
  itCreatesAMultiWordServiceTestFile(baseArgs)
  itCreatesASingleWordServiceFileWithCRUDActions(baseArgs)
  itCreatesASingleWordServiceTestFileWithCRUDActions(baseArgs)
  itCreatesAMultiWordServiceFileWithCRUDActions(baseArgs)
  itCreatesAMultiWordServiceTestFileWithCRUDActions(baseArgs)
  itCreatesASingleWordServiceFileWithAHasManyRelation(baseArgs)
  itCreatesASingleWordServiceFileWithABelongsToRelation(baseArgs)
  itCreatesASingleWordServiceFileWithMultipleRelations(baseArgs)
  itCreatesAMultiWordServiceTestFileWithCRUDAndOnlyForeignKeyRequired(baseArgs)
})

describe('parseSchema', () => {
  it('returns an empty array for models with no required scalars', async () => {
    const { scalarFields } = await service.parseSchema('Product')

    expect(scalarFields).toEqual([])
  })

  it('includes required scalar fields', async () => {
    const { scalarFields } = await service.parseSchema('User')

    expect(
      scalarFields.find((field) => field.name === 'email')
    ).not.toBeUndefined()
  })

  it('does not include non-required scalar fields', async () => {
    const { scalarFields } = await service.parseSchema('User')

    expect(scalarFields.find((field) => field.name === 'name')).toBeUndefined()
  })

  it('does not include required scalar fields with default values', async () => {
    const { scalarFields } = await service.parseSchema('User')

    expect(
      scalarFields.find((field) => field.name === 'isAdmin')
    ).toBeUndefined()
  })

  it('includes foreign key scalars', async () => {
    const { scalarFields } = await service.parseSchema('UserProfile')

    expect(
      scalarFields.find((field) => field.name === 'userId')
    ).not.toBeUndefined()
  })

  it('does not include prisma-generated helper fields', async () => {
    const { scalarFields } = await service.parseSchema('UserProfile')

    expect(scalarFields.find((field) => field.name === 'user')).toBeUndefined()
  })

  it('returns an empty object for models with no relations', async () => {
    const { relations } = await service.parseSchema('User')

    expect(relations).toEqual({})
  })

  it('returns relations', async () => {
    const { relations } = await service.parseSchema('UserProfile')

    expect(relations).toEqual({
      user: { foreignKey: ['userId'], type: 'User' },
    })
  })

  it('returns relations for join tables', async () => {
    const { relations } = await service.parseSchema('TagsOnProducts')

    expect(relations).toEqual({
      tag: { foreignKey: ['tagId'], type: 'Tag' },
      product: { foreignKey: ['productId'], type: 'Product' },
    })
  })

  it('properly captures relationships with different field name than related model', async () => {
    const { relations } = await service.parseSchema('Feature')

    expect(relations).toEqual({
      inventory: { foreignKey: ['inventoryId'], type: 'Product' },
    })
  })

  it('returns an empty array for models with no foreign keys', async () => {
    const { foreignKeys } = await service.parseSchema('User')

    expect(foreignKeys).toEqual([])
  })

  it('returns foreign keys', async () => {
    const { foreignKeys } = await service.parseSchema('UserProfile')

    expect(foreignKeys).toEqual(['userId'])
  })
})

describe('fieldsToScenario', () => {
  it('includes scalar fields', async () => {
    const output = await service.fieldsToScenario(
      [{ name: 'email', type: 'String' }],
      {},
      []
    )

    expect(output).toEqual({ email: 'String' })
  })

  it('includes dependent relationships', async () => {
    // fields for a Post model
    const output = await service.fieldsToScenario(
      [
        { name: 'title', type: 'String' },
        { name: 'userId', type: 'Integer' },
      ],
      { user: { foreignKey: 'userId', type: 'User' } },
      ['userId']
    )

    expect(Object.keys(output)).toEqual(['title', 'user'])
    expect(Object.keys(output.user)).toEqual(['create'])
    expect(Object.keys(output.user.create)).toEqual(['email'])
  })

  it('properly looks up related models by type', async () => {
    // fields for a Post model
    const output = await service.fieldsToScenario(
      [
        { name: 'title', type: 'String' },
        { name: 'userId', type: 'Integer' },
      ],
      // note that relationship name is "author" but datatype is "User"
      { author: { foreignKey: 'authorId', type: 'User' } },
      ['userId']
    )

    expect(Object.keys(output)).toEqual(['title', 'author'])
    expect(Object.keys(output.author)).toEqual(['create'])
    expect(Object.keys(output.author.create)).toEqual(['email'])
  })
})

test("doesn't include test file when --tests is set to false", async () => {
  const baseArgs = {
    ...getDefaultArgs(service.defaults),
    javascript: true,
  }

  const files = await service.files({
    ...baseArgs,
    name: 'User',
    tests: false,
  })

  expect(Object.keys(files)).toEqual([
    path.normalize('/path/to/project/api/src/services/users/users.js'),
  ])
})
