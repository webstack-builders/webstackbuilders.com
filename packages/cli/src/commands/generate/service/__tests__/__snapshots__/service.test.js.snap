// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in javascript mode creates a multi word service file 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const userProfiles = () => {
  return db.userProfile.findMany()
}
"
`;

exports[`in javascript mode creates a multi word service test file 1`] = `
"import { userProfiles } from './userProfiles'

describe('userProfiles', () => {
  scenario('returns all userProfiles', async (scenario) => {
    const result = await userProfiles()

    expect(result.length).toEqual(Object.keys(scenario.userProfile).length)
  })
})
"
`;

exports[`in javascript mode creates a multi word service test file with crud actions and only foreign as mandatory field 1`] = `
"import {
  transactions,
  transaction,
  createTransaction,
  updateTransaction,
  deleteTransaction,
} from './transactions'

describe('transactions', () => {
  scenario('returns all transactions', async (scenario) => {
    const result = await transactions()

    expect(result.length).toEqual(Object.keys(scenario.transaction).length)
  })

  scenario('returns a single transaction', async (scenario) => {
    const result = await transaction({ id: scenario.transaction.one.id })

    expect(result).toEqual(scenario.transaction.one)
  })

  scenario('creates a transaction', async (scenario) => {
    const result = await createTransaction({
      input: { userId: scenario.transaction.two.userId },
    })

    expect(result.userId).toEqual(scenario.transaction.two.userId)
  })

  scenario('updates a transaction', async (scenario) => {
    const original = await transaction({ id: scenario.transaction.one.id })
    const result = await updateTransaction({
      id: original.id,
      input: { userId: scenario.transaction.two.userId },
    })

    expect(result.userId).toEqual(scenario.transaction.two.userId)
  })

  scenario('deletes a transaction', async (scenario) => {
    const original = await deleteTransaction({
      id: scenario.transaction.one.id,
    })

    const result = await transaction({ id: original.id })

    expect(result).toEqual(null)
  })
})
"
`;

exports[`in javascript mode creates a single word service file 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}
"
`;

exports[`in javascript mode creates a single word service file with CRUD actions 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const posts = () => {
  return db.post.findMany()
}

export const post = ({ id }) => {
  return db.post.findUnique({
    where: { id },
  })
}

export const createPost = ({ input }) => {
  return db.post.create({
    data: input,
  })
}

export const updatePost = ({ id, input }) => {
  return db.post.update({
    data: input,
    where: { id },
  })
}

export const deletePost = ({ id }) => {
  return db.post.delete({
    where: { id },
  })
}
"
`;

exports[`in javascript mode creates a single word service file with a belongsTo relation 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  identity: (_obj, { root }) =>
    db.user.findUnique({ where: { id: root.id } }).identity(),
}
"
`;

exports[`in javascript mode creates a single word service file with a hasMany relation 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  userProfiles: (_obj, { root }) =>
    db.user.findUnique({ where: { id: root.id } }).userProfiles(),
}
"
`;

exports[`in javascript mode creates a single word service file with multiple relations 1`] = `
"import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  userProfiles: (_obj, { root }) =>
    db.user.findUnique({ where: { id: root.id } }).userProfiles(),
  identity: (_obj, { root }) =>
    db.user.findUnique({ where: { id: root.id } }).identity(),
}
"
`;

exports[`in javascript mode creates a single word service test file 1`] = `
"import { users } from './users'

describe('users', () => {
  scenario('returns all users', async (scenario) => {
    const result = await users()

    expect(result.length).toEqual(Object.keys(scenario.user).length)
  })
})
"
`;

exports[`in typescript mode creates a multi word service file 1`] = `
"import type { BeforeResolverSpecType } from '@redwoodjs/graphql-server'

import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const userProfiles = () => {
  return db.userProfile.findMany()
}
"
`;

exports[`in typescript mode creates a multi word service test file 1`] = `
"import { userProfiles } from './userProfiles'
import type { StandardScenario } from './userProfiles.scenarios'

describe('userProfiles', () => {
  scenario('returns all userProfiles', async (scenario: StandardScenario) => {
    const result = await userProfiles()

    expect(result.length).toEqual(Object.keys(scenario.userProfile).length)
  })
})
"
`;

exports[`in typescript mode creates a multi word service test file with crud actions and only foreign as mandatory field 1`] = `
"import {
  transactions,
  transaction,
  createTransaction,
  updateTransaction,
  deleteTransaction,
} from './transactions'
import type { StandardScenario } from './transactions.scenarios'

describe('transactions', () => {
  scenario('returns all transactions', async (scenario: StandardScenario) => {
    const result = await transactions()

    expect(result.length).toEqual(Object.keys(scenario.transaction).length)
  })

  scenario(
    'returns a single transaction',
    async (scenario: StandardScenario) => {
      const result = await transaction({ id: scenario.transaction.one.id })

      expect(result).toEqual(scenario.transaction.one)
    }
  )

  scenario('creates a transaction', async (scenario: StandardScenario) => {
    const result = await createTransaction({
      input: { userId: scenario.transaction.two.userId },
    })

    expect(result.userId).toEqual(scenario.transaction.two.userId)
  })

  scenario('updates a transaction', async (scenario: StandardScenario) => {
    const original = await transaction({ id: scenario.transaction.one.id })
    const result = await updateTransaction({
      id: original.id,
      input: { userId: scenario.transaction.two.userId },
    })

    expect(result.userId).toEqual(scenario.transaction.two.userId)
  })

  scenario('deletes a transaction', async (scenario: StandardScenario) => {
    const original = await deleteTransaction({
      id: scenario.transaction.one.id,
    })
    const result = await transaction({ id: original.id })

    expect(result).toEqual(null)
  })
})
"
`;

exports[`in typescript mode creates a single word service file 1`] = `
"import type { BeforeResolverSpecType } from '@redwoodjs/graphql-server'

import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}
"
`;

exports[`in typescript mode creates a single word service file with CRUD actions 1`] = `
"import type { Prisma } from '@prisma/client'
import type { BeforeResolverSpecType } from '@redwoodjs/graphql-server'

import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const posts = () => {
  return db.post.findMany()
}

export const post = ({ id }: Prisma.PostWhereUniqueInput) => {
  return db.post.findUnique({
    where: { id },
  })
}

interface CreatePostArgs {
  input: Prisma.PostCreateInput
}

export const createPost = ({ input }: CreatePostArgs) => {
  return db.post.create({
    data: input,
  })
}

interface UpdatePostArgs extends Prisma.PostWhereUniqueInput {
  input: Prisma.PostUpdateInput
}

export const updatePost = ({ id, input }: UpdatePostArgs) => {
  return db.post.update({
    data: input,
    where: { id },
  })
}

export const deletePost = ({ id }: Prisma.PostWhereUniqueInput) => {
  return db.post.delete({
    where: { id },
  })
}
"
`;

exports[`in typescript mode creates a single word service file with a belongsTo relation 1`] = `
"import type { Prisma } from '@prisma/client'
import type {
  ResolverArgs,
  BeforeResolverSpecType,
} from '@redwoodjs/graphql-server'

import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  identity: (_obj, { root }: ResolverArgs<ReturnType<typeof user>>) =>
    db.user.findUnique({ where: { id: root.id } }).identity(),
}
"
`;

exports[`in typescript mode creates a single word service file with a hasMany relation 1`] = `
"import type { Prisma } from '@prisma/client'
import type {
  ResolverArgs,
  BeforeResolverSpecType,
} from '@redwoodjs/graphql-server'

import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  userProfiles: (_obj, { root }: ResolverArgs<ReturnType<typeof user>>) =>
    db.user.findUnique({ where: { id: root.id } }).userProfiles(),
}
"
`;

exports[`in typescript mode creates a single word service file with multiple relations 1`] = `
"import type { Prisma } from '@prisma/client'
import type {
  ResolverArgs,
  BeforeResolverSpecType,
} from '@redwoodjs/graphql-server'

import { db } from 'src/lib/db'
import { requireAuth } from 'src/lib/auth'

// Used when the environment variable REDWOOD_SECURE_SERVICES=1
export const beforeResolver = (rules: BeforeResolverSpecType) => {
  rules.add(requireAuth)
}

export const users = () => {
  return db.user.findMany()
}

export const User = {
  userProfiles: (_obj, { root }: ResolverArgs<ReturnType<typeof user>>) =>
    db.user.findUnique({ where: { id: root.id } }).userProfiles(),
  identity: (_obj, { root }: ResolverArgs<ReturnType<typeof user>>) =>
    db.user.findUnique({ where: { id: root.id } }).identity(),
}
"
`;

exports[`in typescript mode creates a single word service test file 1`] = `
"import { users } from './users'
import type { StandardScenario } from './users.scenarios'

describe('users', () => {
  scenario('returns all users', async (scenario: StandardScenario) => {
    const result = await users()

    expect(result.length).toEqual(Object.keys(scenario.user).length)
  })
})
"
`;
